metaflow/plugins/cards/card_cli.py

```diff
@@ -0,0 +1,388 @@
+
+
+@contextmanager
+def timeout(time):
+    # Register a function to raise a TimeoutError on the signal.
+    signal.signal(signal.SIGALRM, raise_timeout)
+    # Schedule the signal to be sent after ``time``.
+    signal.alarm(time)
+
+    try:
+        yield
+    except TimeoutError:
+        pass
+    finally:
+        # Unregister the signal so it won't be triggered
+        # if the timeout is not reached.
+        signal.signal(signal.SIGALRM, signal.SIG_IGN)
+
+
+def raise_timeout(signum, frame):
+    raise TimeoutError
+
+
+def list_available_cards(ctx, path_spec, card_paths, card_datastore, command="view"):
+    # todo : create nice response messages on the CLI for cards which were found.
+    scriptname = ctx.obj.flow.script_name
+    path_tuples = card_datastore.get_card_names(card_paths)
+    ctx.obj.echo(
+        "\nFound %d card matching for your query..." % len(path_tuples), fg="green"
+    )
+    task_pathspec = "/".join(path_spec.split("/")[1:])
+    card_list = []
+    for path_tuple in path_tuples:
+        card_name = "%s-%s" % (path_tuple.type, path_tuple.hash)
+        card_list.append(card_name)
+
+    random_idx = 0 if len(path_tuples) == 1 else random.randint(0, len(path_tuples) - 1)
+    _, randhash = path_tuples[random_idx]
+    ctx.obj.echo("\n\t".join([""] + card_list), fg="blue")
+    ctx.obj.echo(
+        "\n\tExample access from CLI via: \n\t %s\n"
+        % make_command(
+            scriptname,
+            task_pathspec,
+            command=command,
+            hash=randhash[:NUM_SHORT_HASH_CHARS],
+        ),
+        fg="yellow",
+    )
+
+
+def make_command(
+    script_name,
+    taskspec,
+    command="get",
+    hash=None,
+):
+    calling_args = ["--hash", hash]
+    return " ".join(
+        [
+            ">>>",
+            "python",
+            script_name,
+            "card",
+            command,
+            taskspec,
+        ]
+        + calling_args
+    )
+
+
+@click.group()
+def cli():
+    pass
+
+
+@cli.group(help="Commands related to @card decorator.")
+@click.pass_context
+def card(ctx):
+    # We set the metadata values here so that top level arguments to --datastore and --metadata
+    # Can work with the Metaflow client.
+    # If we don't set the metadata here than the metaflow client picks the defaults when calling the `Task`/`Run` objects. These defaults can come from the `config.json` file or based on the `METAFLOW_PROFILE`
+    from metaflow import metadata
+
+    setting_metadata = "@".join(
+        [ctx.obj.metadata.TYPE, ctx.obj.metadata.default_info()]
+    )
+    metadata(setting_metadata)
+    # set the card root to the datastore according to the configuration.
+    root_pth = CardDatastore.get_storage_root(ctx.obj.flow_datastore._storage_impl.TYPE)
+    if root_pth is not None:
+        ctx.obj.flow_datastore._storage_impl.datastore_root = root_pth
+
+
+def card_read_options_and_arguments(func):
+    @click.option(
```

Comment:
We can expose this but since there is only one card, maybe consider removing for now.

Reply:
can you elaborate ?

Reply:
If there is only one type of card, and assuming that generating the card for a step is stable, there should only be one hash ever available. It is fine to leave though since it won't be much advertised I guess.

Reply:
Cards can be created from the cli using `card create` command. Cards also support `options` that can account for creating cards of a different hash altogether ( when created from CLI ). Due to this, we have the hash.